test/notifications/open_test.dart

```diff
@@ -0,0 +1,577 @@
+      takeStartingRoutes(account: eg.selfAccount);
+      // … and then the one the notification leads to.
+      matchesNavigation(check(pushedRoutes).single, eg.selfAccount, message);
+    }, variant: const TargetPlatformVariant({TargetPlatform.android, TargetPlatform.iOS}));
+
+    testWidgets('at app launch', (tester) async {
+      addTearDown(testBinding.reset);
+      final account = eg.selfAccount;
+      final message = eg.streamMessage();
+      setupNotificationDataForLaunch(tester, account, message);
+
+      // Now start the app.
+      await prepare(tester, dropStartingRoutes: false);
+      check(pushedRoutes).isEmpty(); // GlobalStore hasn't loaded yet
+
+      // Once the app is ready, we navigate to the conversation.
+      await tester.pump();
+      takeStartingRoutes(account: account);
+      matchesNavigation(check(pushedRoutes).single, account, message);
+    }, variant: const TargetPlatformVariant({TargetPlatform.android, TargetPlatform.iOS}));
+
+    testWidgets('uses associated account as initial account, if initial route', (tester) async {
+      addTearDown(testBinding.reset);
+
+      final accountA = eg.selfAccount;
+      final accountB = eg.otherAccount;
+      final message = eg.streamMessage();
+      await testBinding.globalStore.add(accountA, eg.initialSnapshot());
+      await testBinding.globalStore.add(accountB, eg.initialSnapshot());
+      setupNotificationDataForLaunch(tester, accountB, message);
+
+      await prepare(tester, dropStartingRoutes: false, withAccount: false);
+      check(pushedRoutes).isEmpty(); // GlobalStore hasn't loaded yet
+
+      await tester.pump();
+      takeStartingRoutes(account: accountB);
+      matchesNavigation(check(pushedRoutes).single, accountB, message);
+    }, variant: const TargetPlatformVariant({TargetPlatform.android, TargetPlatform.iOS}));
+  });
+
+  group('NotificationOpenPayload', () {
+    test('android: smoke round-trip', () {
+      // DM narrow
+      var payload = NotificationOpenPayload(
+        realmUrl: Uri.parse('http://chat.example'),
+        userId: 1001,
+        narrow: DmNarrow(allRecipientIds: [1001, 1002], selfUserId: 1001),
+      );
+      var url = payload.buildAndroidNotificationUrl();
+      check(NotificationOpenPayload.parseAndroidNotificationUrl(url))
+        ..realmUrl.equals(payload.realmUrl)
+        ..userId.equals(payload.userId)
+        ..narrow.equals(payload.narrow);
+
+      // Topic narrow
+      payload = NotificationOpenPayload(
+        realmUrl: Uri.parse('http://chat.example'),
+        userId: 1001,
+        narrow: eg.topicNarrow(1, 'topic A'),
+      );
+      url = payload.buildAndroidNotificationUrl();
+      check(NotificationOpenPayload.parseAndroidNotificationUrl(url))
+        ..realmUrl.equals(payload.realmUrl)
+        ..userId.equals(payload.userId)
+        ..narrow.equals(payload.narrow);
+    });
+
+    group('parseIosApnsPayload', () {
+      test('smoke one-one DM', () {
```

Comment:
Cool, these tests are a good start.

Because of the nature of this function as parsing code (echoing https://github.com/zulip/zulip-flutter/pull/1379#discussion_r2099235693), and because notifications are important, I'd like to have more thorough testing for it than only smoke tests.

But let's make that a follow-up issue we can pursue after launch. Even if there are edge cases this code doesn't do anything on, that won't be a regression from what's in main. (And if there are edge cases where it does something but it's the wrong thing, going to the wrong narrow is probably not much worse than doing nothing.)

Reply:
Filed: https://github.com/zulip/zulip-flutter/issues/1563